abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,919 @@
+
+* params:
+  - `i32 (const char*) queue_name_data`
+  - `i32 (size_t) queue_name_size`
+  - `i32 (uint32_t*) return_queue_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Resolves existing shared queue using a given name (`queue_name_data`, `queue_name_size`). It can be
+referred to in `proxy_enqueue_shared_queue` and `proxy_dequeue_shared_queue` using returned unique
+queue identifier (`queue_id`).
+
+
+### `proxy_dequeue_shared_queue`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+  - `i32 (const char**) payload_data`
+  - `i32 (size_t*) payload_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Get data (`payload_data`, `payload_size`) to the end of the queue (`queue_id`).
+
+
+### `proxy_enqueue_shared_queue`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+  - `i32 (const char*) payload_data`
+  - `i32 (size_t) payload_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Add data (`payload_data`, `payload_size`) to the front of the queue (`queue_id`).
+
+
+### `proxy_remove_shared_queue`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Remove queue (`queue_id`).
+
+
+## Stats/metrics
+
+### `proxy_define_metric`
+
+* params:
+  - `i32 (proxy_metric_type_t) metric_type`
+  - `i32 (const char*) metric_name_data`
+  - `i32 (size_t) metric_name_size`
+  - `i32 (uint32_t*) return_metric_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Define a metric using a given name (`metric_name_data`, `metic_name_size`). It can be referred to in
+`proxy_get_metric`, `proxy_increment_metric` and `proxy_record_metric` using returned unique metric
+identifier (`metric_id`).
+
+
+### `proxy_get_metric`
+
+* params:
+  - `i32 (uint32_t) metric_id`
+  - `i32 (uint64_t*) return_value`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Get the value of the metric (`metric_id`).
+
+
+### `proxy_record_metric`
+
+* params:
+  - `i32 (uint32_t) metric_id`
+  - `i64 (uint64_t) value
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Set the value of the metric (`metric_id`) to the value (`value`).
+
+
+### `proxy_increment_metric`
+
+* params:
+  - `i32 (uint32_t) metric_id`
+  - `i64 (int64_t) offset
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Increment/decrement value of the metric (`metric_id`) by offset (`offset`).
+
+
+### `proxy_remove_metric`
+
+* params:
+  - `i32 (uint32_t) metric_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Remove the metric (`metric_id`).
+
+
+## Foreign function interface (FFI):
+
+### `proxy_call_foreign_function`
+
+* params:
+  - `i32 (const char*) function_name_data`
+  - `i32 (size_t) function_name_size`
+  - `i32 (const char *) parameters_data`
+  - `i32 (size_t) parameters_size`
+  - `i32 (const char**) return_results_data`
+  - `i32 (size_t*) return_results_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Call registered foreign function (`function_name_data`, `function_name_size`).
```

Comment:
What is this magical device?

Reply:
Foreign function interface (FFI). Basically, it allows host to register custom functions, that are not part of this ABI, and that will be accessible to the Wasm VM.

Reply:
Add note: return_results_data and return_results_size can be nullptr if the function is not expected to return anything.

Reply:
We need an "on_foreign_function" call into the VM, passing a function_id and a data_size.  This will require another BufferType for retrieving the data.